### PR TITLE
Refactor some tests to assert on change events instead of internal state

### PR DESCRIPTION
**Pull request by jimmycallin**
Refactored a few test fields to assert on change events instead of internal state.

This is just the beginning :) If you like this I'll refactor the rest as well.

### Reasons for making this change

[Please describe them here]

If this is related to existing tickets, include links to them as well.

### Checklist

* [ ] **I'm updating documentation**
  - [ ] I've [checked the rendering](https://react-jsonschema-form.readthedocs.io/en/latest/#contributing) of the Markdown text I've added
* [x] **I'm adding or updating code**
  - [x] I've added and/or updated tests
  - [ ] I've updated [docs](https://react-jsonschema-form.readthedocs.io/) if needed
* [ ] **I'm adding a new feature**
  - [ ] I've updated the playground with an example use of the feature

### CODE REVIEW

File: packages/core/test/test_utils.js [hunk 1]
```diff
@@ -43,3 +53,9 @@ export function describeRepeated(title, fn) {
     );
   }
 }
+
+export async function submitForm(node) {
```
Comment (epicfaace) [inline packages/core/test/test_utils.js]: Why does this function (and the function wrapped in `act`) need to be async?

File: packages/core/test/NumberField_test.js [hunk 2]
```diff
@@ -117,13 +122,15 @@ describe("NumberField", () => {
         expect(node.querySelector(".field-description").textContent).eql("bar");
       });
 
-      it("should default state value to undefined", () => {
-        const { comp } = createFormComponent({
+      it("should default state value to undefined", async () => {
```
Comment (epicfaace) [inline packages/core/test/NumberField_test.js]: Can you change the test name here?

**Review by epicfaace — COMMENTED** [review]
Thanks, looks good!
- When would one want to test `onChange` versus `onSubmit`? I don't see a clear distinction between when you decided to test each one in the tests you added.
- Could we replace something like `onChange.lastCall.args[0].formData` with the less verbose `onChange.calledOnceWithExactly({})`? This is a great way to refactor the tests, but it might be good to figure out what our style of assertions will look like so we can standardize across all tests.

File: packages/core/test/test_utils.js (hunk 1 above)
Comment (jimmycallin) [reply packages/core/test/test_utils.js]: apparently it didn't :) still haven't quite figured out when and when not you need `act` and when to make it async.

**Comment by jimmycallin** [comment]
`calledOnceWithExactly` became a bit too verbose imo since you needed to pass a bunch of for the test unrelated parameters. took a middle road with `calledWithMatch` to only assert that formData is set correctly.

**Comment by jimmycallin** [comment]
> When would one want to test onChange versus onSubmit? I don't see a clear distinction between when you decided to test each one in the tests you added.

it seems like `onChange` is not called when you pass `formData`, but when you set default values  `onChange` is indeed called. the tests should (now) reflect that `submitForm` is only called when you pass formData on initiation and want to verify that it is set correctly.

**Comment by epicfaace** [comment]
> `calledOnceWithExactly` became a bit too verbose imo since you needed to pass a bunch of for the test unrelated parameters. took a middle road with `calledWithMatch` to only assert that formData is set correctly.

Good idea. Maybe you could replace `sinon.assert.calledWithMatch` with `spy.calledWithMatch` to make it even less verbose?

**Comment by jimmycallin** [comment]
> Good idea. Maybe you could replace sinon.assert.calledWithMatch with spy.calledWithMatch to make it even less verbose?

the output of `spy.calledWithMatch` is only `true` or `false`, which means you have to wrap it in expect(...).eql(...). in case of mismatch, the error message is very unhelpful (e.g. "true should be false"), while I found using sinon's built in assertion much more helpful in pinpointing the error. but if you insist on using `expect(spy.calledWithMatch(...)).eql(...)` i'll change it :)

**Comment by epicfaace** [comment]
Ok, that makes sense -- since sinon's assertion error messages are better than using `expect` and `eql`, I'm all for it.

**Comment by epicfaace** [comment]
> it seems like `onChange` is not called when you pass `formData`, but when you set default values `onChange` is indeed called.

Note, however, that we'll probably change this behavior in v2 so that `onChange` is not called when setting default values (#1201).

**Review by epicfaace — APPROVED** [review]
Looks good -- thanks! It'd be great if you are able to help refactor the rest as well.

**Review by epicfaace — APPROVED** [review]
Changes look good, thanks! Thanks for making several of the tests more robust as well.

**Pull request merged**